commit c693639dd2722dbb455ae824994bcb997e0fa6e5 (refs/changes/60/1460/5)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-09-05T22:24:50+00:00 (2 years, 1 month ago)
    
    MORAY-389 Limit which modules triggers are allowed to require
    Reviewed by: David Pacheco <dap@joyent.com>
    Reviewed by: Patrick Mooney <patrick.mooney@joyent.com>
    Approved by: Joshua M. Clulow <jmc@joyent.com>

diff --git a/test/buckets.test.js b/test/buckets.test.js
index b20ca90..c624f39 100644
--- a/test/buckets.test.js
+++ b/test/buckets.test.js
@@ -677,3 +677,68 @@ test('MORAY-388 - triggers can require microtime/crc modules', function (t) {
         t.end();
     });
 });
+
+
+test('MORAY-389 - limit what triggers can require', function (t) {
+    function requireAssertPlus(req, cb) {
+        try {
+            require('assert-plus');
+            cb(new Error('should have thrown'));
+        } catch (e) {
+            cb(e);
+        }
+    }
+
+    function requireNet(req, cb) {
+        try {
+            require('net');
+            cb(new Error('should have thrown'));
+        } catch (e) {
+            cb(e);
+        }
+    }
+
+    function testTrigger(name, cb) {
+        c.putObject(b, 'key1', {}, function (err) {
+            if (err) {
+                var cause = VError.findCauseByName(err, 'InvalidRequireError');
+                if (cause) {
+                    t.equal(cause.message,
+                        '"' + name + '" is not a permitted module');
+                } else {
+                    t.error(err, name + ' module was required');
+                }
+            } else {
+                t.fail('trigger not called');
+            }
+
+            cb();
+        });
+    }
+
+    var schema1 = { pre: [ requireAssertPlus ] };
+    var schema2 = { post: [ requireAssertPlus ] };
+    var schema3 = { pre: [ requireNet ] };
+    var schema4 = { post: [ requireNet ] };
+
+    vasync.pipeline({ funcs: [
+        // Create the initial bucket to check "pre" assert-plus trigger.
+        function (_, cb) { c.createBucket(b, schema1, cb); },
+        function (_, cb) { testTrigger('assert-plus', cb); },
+
+        // Update the bucket to check "post" assert-plus trigger.
+        function (_, cb) { c.updateBucket(b, schema2, cb); },
+        function (_, cb) { testTrigger('assert-plus', cb); },
+
+        // Update the bucket to check "post" net trigger.
+        function (_, cb) { c.updateBucket(b, schema3, cb); },
+        function (_, cb) { testTrigger('net', cb); },
+
+        // Update the bucket to check "post" net trigger.
+        function (_, cb) { c.updateBucket(b, schema4, cb); },
+        function (_, cb) { testTrigger('net', cb); }
+    ]}, function (err) {
+        t.error(err, 'Finish without error');
+        t.end();
+    });
+});
